src/modules/item/sagas.ts

```diff
@@ -200,43 +208,65 @@ export function* itemSaga(legacyBuilder: LegacyBuilderAPI, builder: BuilderClien
     }
   }
 
+  function* handleCreateOrEditProgress(action: { progress: number }) {
+    yield put(updateProgressSaveMultipleItems(action.progress))
+  }
+
+  function* handleCreateOrEditCancelledItems(action: SaveMultipleItemsCancelledAction['payload']) {
+    const { items, savedFileNames, notSavedFileNames, cancelledFileNames } = action
+    yield put(saveMultipleItemsCancelled(items, savedFileNames, notSavedFileNames, cancelledFileNames))
+  }
+
   function* handleSaveMultipleItemsRequest(action: SaveMultipleItemsRequestAction) {
     const { builtFiles } = action.payload
+    const fileNamesSucceeded: string[] = []
+    const fileNamesFailed: string[] = []
     const remoteItems: RemoteItem[] = []
-    const fileNames: string[] = []
-
-    // Upload files sequentially to avoid DoSing the server
+    const queue = new PQueue({ concurrency: SAVE_AND_EDIT_FILES_BATCH_SIZE })
     try {
-      for (const [index, builtFile] of builtFiles.entries()) {
-        const remoteItem: RemoteItem = yield call([builder, 'upsertItem'], builtFile.item, builtFile.newContent)
-        remoteItems.push(remoteItem)
-        fileNames.push(builtFile.fileName)
-        yield put(updateProgressSaveMultipleItems(Math.round(((index + 1) / builtFiles.length) * 100)))
+      const promisesOfItemsToSave: (() => Promise<void>)[] = []
+
+      for (const [_index, builtFile] of builtFiles.entries()) {
+        promisesOfItemsToSave.push(async () => {
+          try {
+            const remoteItem: RemoteItem = await builder.upsertItem(builtFile.item, builtFile.newContent)
+            fileNamesSucceeded.push(builtFile.fileName)
+            remoteItems.push(remoteItem)
+          } catch (error) {
+            fileNamesFailed.push(builtFile.fileName)
+          }
+        })
       }
 
+      queue.on('next', () => {
+        createOrEditProgressChannel.put({ progress: Math.round(((builtFiles.length - queue.pending) / builtFiles.length) * 100) })
+      })
+
+      yield queue.addAll(promisesOfItemsToSave)
+
       yield put(
         saveMultipleItemsSuccess(
           remoteItems.map(remoteItem => fromRemoteItem(remoteItem)),
-          fileNames
-        )
-      )
-    } catch (error) {
-      yield put(
-        saveMultipleItemsFailure(
-          error.message,
-          remoteItems.map(remoteItem => fromRemoteItem(remoteItem)),
-          fileNames
+          fileNamesSucceeded,
+          fileNamesFailed
         )
       )
     } finally {
       const wasCancelled: boolean = yield cancelled()
       if (wasCancelled) {
-        yield put(
-          saveMultipleItemsCancelled(
-            remoteItems.map(remoteItem => fromRemoteItem(remoteItem)),
-            fileNames
+        queue.clear()
```

Comment:
I believe that we ned to `clear` the queue since pausing it will just put the execution on hold, usually to be resumed later on. In our case, we won't resume it so I guess we just should get rid of the unprocessed ones.